### DISCUSSION TITLE
Python part of Ariadne in WALA?

### DISCUSSION THREAD

**Discussion by khatchad** [wala/WALA#1412] [Ideas]
We may be getting to a point now where the Python parsing part of Ariadne (`com.ibm.wala.cast.python`) can join Java and Javascript in the `cast` module on WALA. The ML part of Ariadne should, I believe, remain in the original repo for now until it is more mature to handle modern versions of TensorFlow more fully.

Is there any desire for this? My thinking is that,  in doing so, it will be become more accessible to WALA users. Currently, being in its own repository completely may make it a bit obscure. Moving it would also potentially make people aware that it has a lot of new features since it's inception. Specifically, perhaps people tried it before without much luck on some modern Python.

**Comment by msridhar** [comment]
I would be very open to moving the Python frontend to the main WALA repo.  One concern that comes to mind is tests.  Does the frontend itself have decent test coverage that is independent of the ML usage in Ariadne?  I think we would want to have a good test suite that has assertions about correctness of call graph construction, etc.

**Reply by khatchad** [reply]
Yes, this was one of my concerns as well. There is quite a bit of testing outside of the ML usage for the original features, but perhaps we leaned on the ML testing to test the new features. 

I can run a coverage tool on the frontend tests to get a better sense of this.

**Reply by khatchad** [reply]
One feature specifically that we learned quite a bit on the ML infrastructure was to test the functors (callable objects) as this is heavily used in ML code. We might have to write some additional tests outside of the ML suite for features like these.

**Reply by XC-Xinze** [reply]
We ran the coverage tool and found out that by removing the ML tests (`com.ibm.wala.cast.python.ml.tests`), we lose 5% instruction coverage and 7% in branch coverage for `com.ibm.wala.cast.python.ml.tests`. Specifically, we go from 55% code coverage down to 50% and 42% branch coverage down to 35%:

## Original

Element | Missed Instructions | Cov. | Missed Branches | Cov. | Missed | Cxty | Missed | Lines | Missed | Methods | Missed | Classes
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
com.ibm.wala.cast.python |   | 55% |   | 42% | 890 | 1,467 | 1,622 | 3,684 | 474 | 850 | 46 | 116

## Removing ML Tests

Element | Missed Instructions | Cov. | Missed Branches | Cov. | Missed | Cxty | Missed | Lines | Missed | Methods | Missed | Classes
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
com.ibm.wala.cast.python |   | 50% |   | 35% | 950 | 1,467 | 1,837 | 3,684 | 493 | 850 | 47 | 116

**Reply by khatchad** [reply]
This is great, @XC-Xinze , thank you. Can you please compare the two runs to see exactly which code coverage we are losing in `com.ibm.wala.cast.python` between the two runs? If we can understand the code coverage we are losing, perhaps we can add some tests to get them even.

**Reply by XC-Xinze** [reply]
Attaching reports:
- [original.zip](https://github.com/user-attachments/files/20211661/original.zip)
- [removed.zip](https://github.com/user-attachments/files/20211648/removed.zip)

**Reply by khatchad** [reply]
I sat down with @XC-Xinze today to have a look at the methods and classes that lose coverage after removing the ML tests (he'll post the details soon). The most glaring loss has to do with Python packages, which is a new feature. Unfortunately, I don;t think this will be easy to add to the normal CAst tests as there was quite a bit of infrastructure needed in the ML tests to support these.

**Reply by juliandolby** [reply]
> I would be very open to moving the Python frontend to the main WALA repo. One concern that comes to mind is tests. Does the frontend itself have decent test coverage that is independent of the ML usage in Ariadne? I think we would want to have a good test suite that has assertions about correctness of call graph construction, etc.

The tests for Python focus on individual features, but, to make life simpler, they are all designed to impact the call graph, and we check that the expected call graph nodes are there.  Thus every test is assertions about call graph correctness.

**Reply by khatchad** [reply]
I think that the problem is that, to support additional Python features (e.g., Python packages), we added some code to the classes in the `com.ibm.wala.cast.python` package. The tests for *some* of that code, as @msridhar suspected correctly, is located in `com.ibm.wala.cast.python.ml.tests`. Currently, separating the packages would unfortunately lose some test coverage. There's quite a bit of test infrastructure in `com.ibm.wala.cast.python.ml.tests` to test Python modules that are contained in packages. We would need to replicate that kind of infrastructure (e.g., adding `PYTHONPATH`) to `com.ibm.wala.cast.python.tests` to recover the lost coverage that would result from the separation.

**Reply by msridhar** [reply]
I guess the significant task in moving over that testing logic is writing new assertions just related to the frontend but not to ML stuff?  I don't want the perfect to be the enemy of the good here.  We could do a migration and then gradually add more test coverage directly to the WALA repo, for example.

**Reply by XC-Xinze** [reply]
<google-sheets-html-origin>
Subpackage belongs Package: com.ibm.wala.cast.python | CLASS belong current subpackage | METHOD belong current class | Original Missed Instructions Cov. | Removed Missed Instructions Cov. | Diff between Original and Removed | Original Missed Lines | Removed Missed Lines | Diff between Original and Removed | All Lines
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
com.ibm.wala.cast.python.ipa.callgraph | PytestEntrypointBuilder | isPytestCase(IClass) | 92% | 88% | 4% | 0 | 0 | 0 | 25
  | PythonClassMethodTrampolineTargetSelector | populate(PythonSummary, int, IClass, PythonInvokeInstruction, Logger) | 87% | 0% | 87% | 3 | 33 | -30 | 33
  |   | shouldProcess(CGNode, CallSiteReference, IClass) | 100% | 62% | 38% | 0 | 0 | 0 | 9
  | PythonConstructorTargetSelector | getCalleeTarget(CGNode, CallSiteReference, IClass) | 99% | 99% | 0% | 0 | 0 | 0 | 95
  | PythonInstanceMethodTrampolineTargetSelector | getCallable(CGNode, IClassHierarchy, PythonInvokeInstruction) | 95% | 73% | 22% | 0 | 9 | -9 | 34
  |   | getCalleeTarget(CGNode, CallSiteReference, IClass) | 100% | 93% | 7% | 0 | 0 | 0 | 7
  |   | populate(PythonSummary, int, IClass, PythonInvokeInstruction, Logger) | 100% | 79% | 21% | 0 | 14 | -14 | 60
  | PythonSSAPropagationCallGraphBuilder.PythonConstraintVisitor | getScriptName(TypeName) | 100% | 78% | 22% | 0 | 1 | -1 | 9
  |   | lambda$processWildcardImports$1(MethodReference, String, Deque) | 68% | 0% | 68% | 2 | 6 | -4 | 6
  |   | lambda$processWildcardImports$2(MethodReference, String, Deque) | 100% | 68% | 32% | 0 | 2 | -2 | 6
  |   | processWildcardImports(AstPropertyRead) | 82% | 55% | 27% | 2 | 12 | -10 | 33
  |   | processWildcardImports(SSAInstruction, String, String) | 92% | 86% | 6% | 2 | 4 | -2 | 39
  |   | visitPropertyRead(AstPropertyRead) | 91% | 82% | 9% | 2 | 4 | -2 | 20
  | PythonSSAPropagationCallGraphBuilder.PythonConstraintVisitor.new PythonInstructionVisitor() {...} | visitPut(SSAPutInstruction) | 100% | 28% | 72% | 0 | 4 | -4 | 7
com.ibm.wala.cast.python.ir | PythonCAstToIRTranslator | doPrimitive(int, AstTranslator.WalkContext, CAstNode) | 100% | 83% | 17% | 0 | 10 | -10 | 47
  |   | getLocalModules(List) | 100% | 0% | 100% | 0 | 5 | -5 | 5
  |   | handleObjectLiteralAssign(CAstNode, CAstNode, int, AstTranslator.WalkContext, CAstVisitor) | 100% | 67% | 33% | 0 | 8 | -8 | 27
  |   | isModuleInitializationFile(Path) | 100% | 0% | 100% | 0 | 1 | -1 | 1
  |   | lambda$getLocalModules$9(Set, Iterator) | 100% | 0% | 100% | 0 | 1 | -1 | 1
  |   | lambda$visitScriptEntity$6(String, SourceModule) | 100% | 0% | 100% | 0 | 15 | -15 | 15
  |   | lambda$visitScriptEntity$7(PythonLoader, String, AstTranslator.WalkContext, SourceModule) | 92% | 0% | 92% | 4 | 46 | -42 | 46
  |   | lambda$visitScriptEntity$8(AstTranslator.WalkContext, SSAInstruction) | 100% | 0% | 100% | 0 | 1 | -1 | 1
  |   | leaveVarAssignOp(CAstNode, CAstNode, CAstNode, boolean, AstTranslator.WalkContext, CAstVisitor) | 75% | 0% | 75% | 5 | 26 | -21 | 26
  |   | visitScriptEntity(CAstEntity, AstTranslator.WalkContext, AstTranslator.WalkContext, CAstVisitor) | 100% | 29% | 71% | 0 | 12 | -12 | 17
  |   | visitVarAssignOp(CAstNode, CAstNode, CAstNode, boolean, AstTranslator.WalkContext, CAstVisitor) | 100% | 0% | 100% | 0 | 1 | -1 | 1
  | PythonCAstToIRTranslator.new Consumer() {...} | accept(ModuleEntry) | 58% | 0% | 58% | 1 | 4 | -3 | 4
  |   | {...} | 100% | 0% | 100% | 0 | 1 | -1 | 1
com.ibm.wala.cast.python.ssa | PythonInvokeInstruction | getUse(String) | 100% | 92% | 8% | 0 | 1 | -1 | 4
  | PythonPropertyRead | getExceptionTypes() | 100% | 0% | 100% | 0 | 1 | -1 | 1
com.ibm.wala.cast.python.types | Util | getDeclaringClassTypeReference(TypeReference) | 100% | 0% | 100% | 0 | 4 | -4 | 4
com.ibm.wala.cast.python.util | PythonInterpreter | interpretAsInt(String) | 100% | 0% | 100% | 0 | 1 | -1 | 1
  | Util | getAllocationSiteInNode(ConstantKey) | 65% | 0% | 65% | 1 | 5 | -4 | 5
  |   | getAllocationSiteInNode(InstanceKey) | 29% | 10% | 19% | 10 | 13 | -3 | 14
  |   | isClassMethod(IClass) | 100% | 84% | 16% | 0 | 0 | 0 | 6
  |   | removeFileProtocolFromPath(String) | 100% | 0% | 100% | 0 | 1 | -1 | 1

We compared the coverage of the original file and the version after removing `com.ibm.wala.cast.python.ml.tests`, and sorted out the instruction coverage of the bytecode and the number of original code missing (keyword 'Lines'. The "Original" keyword in the table refers to the original file; the "Removed" keyword refers to the version after the test file is removed.

**Reply by khatchad** [reply]
Thanks, @XC-Xinze. I suppose it doesn't make sense to do the migration now but rather after the transition to Python ASTs.

**Comment by juliandolby** [comment]
With the latest work on using the real Python VM rather than Jython, one option would be to have that and the core Python projects as part of WALA.  The would require people to have a Python VM, but it would make it unnecessary to have all the hard-to-build stuff that Jython requires.

**Reply by khatchad** [reply]
It might be a bit awkward for people not dealing with Python to have a Python VM to build WALA. However, I believe there is a somewhat similar situation with one of the WALA modules that deals with Android (Dalvik?) that has required the Android SDK.

**Reply by msridhar** [reply]
I'm open to anything that integrates reasonably cleanly with WALA's Gradle build system.  But I don't know tricky this will be, particularly if Python VM needs to be installed globally, platform-specific issues, etc.  With Android SDK there is no need for a global install, and all platforms (windows / mac / linux) are supported uniformly, so it's not too bad.

**Reply by khatchad** [reply]
Maybe we can use a [build variant](https://docs.gradle.org/current/userguide/variant_aware_resolution.html) for the Python stuff so that you only need a Python VM when you are building that "variant?"

**Reply by msridhar** [reply]
I think if the stuff is even isolated to the proposed Python sub-project of WALA that is fine.  But it should all work on CI at least on Linux and Mac, and ideally it could work locally without doing any kind of global installation first.

**Comment by khatchad** [comment]
Now that @juliandolby has completed the JEP integration, do we want to revisit this? We had it on hold as it wouldn't have made much sense to port over the Jython stuff and then have to redo the changes for JEP.
